python/lsst/daf/butler/registry/_remoteRegistry.py

```diff
@@ -0,0 +1,431 @@
+
+        self._dimensions: Optional[DimensionUniverse] = None
+
+        # Does each API need to be sent the defaults so that the server
+        # can use specific defaults each time?
+
+        # Storage class information should be pulled from server.
+        # Dimensions should be pulled from server.
+
+    def __str__(self) -> str:
+        return str(self._db)
+
+    def __repr__(self) -> str:
+        return f"RemoteRegistry({self._db!r}, {self.dimensions!r})"
+
+    def isWriteable(self) -> bool:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        # Can be used to prevent any PUTs to server
+        return self._writeable
+
+    def copy(self, defaults: Optional[RegistryDefaults] = None) -> Registry:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        if defaults is None:
+            # No need to copy, because `RegistryDefaults` is immutable; we
+            # effectively copy on write.
+            defaults = self.defaults
+        return type(self)(self._db, defaults, self.isWriteable())
+
+    @property
+    def dimensions(self) -> DimensionUniverse:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        if self._dimensions is not None:
+            return self._dimensions
+
+        # Access /dimensions.json on server and cache it locally.
+        try:
+            response = httpx.get(str(self._db.join("universe")))
+            response.raise_for_status()
+        except httpx.RequestError as e:
+            print(f"An error occurred while requesting {e.request.url!r}.")
+            raise e
+        except httpx.HTTPStatusError as e:
+            print(f"Error response {e.response.status_code} while requesting {e.request.url!r}.")
+            raise e
+
+        config = DimensionConfig.fromString(response.text, format="json")
+        self._dimensions = DimensionUniverse(config)
+        return self._dimensions
+
+    def refresh(self) -> None:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+
+        # Need to determine what to refresh.
+        # Might need a server method to return all the DatasetTypes up front.
+        pass
+
+    # def transaction():
+    #    What does a transaction mean for the client? The server is running
+    #    in a transaction but if the client wants one transaction that
+    #    involves two server methods how does that work? The server will have
+    #    already closed the transaction.
+    @contextlib.contextmanager
+    def transaction(self, *, savepoint: bool = False) -> Iterator[None]:
+        # No-op for now
+        try:
+            yield
+        except BaseException:
+            raise
+
+    # insertOpaqueData + fetchOpaqueData + deleteOpaqueData
+    #    There are no managers for opaque data in client. This implies
+    #    that the server would have to have specific implementations for
+    #    use by Datastore. DatastoreBridgeManager also is not needed.
```

Comment:
I did have designs on being able to use the opaque tables for other things (e.g. backing a `SqlDatastore`'s actual dataset storage, not just its extra records, and using that to save SQuaSH metrics), but I'm not sure yet whether any of them are viable in the no-shared-DB execution environment.  So, yes, let's not add those methods here now, but I'd rather not rule out having something like them in the future.

Reply:
I think I'm edging towards having remote datastore registry provide higher level interfaces but it may well be that this interface does eventually turn up directly on server.